### DISCUSSION TITLE
visually select jsx html tags

### DISCUSSION THREAD

**Discussion by ShooTeX** [LazyVim/LazyVim#500] [Q&A]
`vat` doesn't work correctly in some cases, is there something I can configure in mini.ai to make this work?

https://user-images.githubusercontent.com/10850738/228375260-f3a39f16-36f2-43bb-a879-7ec850851b1e.mov

**Comment by dpetka2001** [comment]
What exactly are you trying to select? Because from the video it looks to me to be doing the correct thing. `vat` will visually select the tags around the content that your cursor currently resides. In your case (according to video) your cursor resides in a `<div>` label nested inside another `<div>` label. So it correctly selects everything inside the `<div>` element that is considered the outside one, including the outbound tags of the outside element.

**Reply by dpetka2001** [reply]
So, I experimented a bit and something I noticed is that it will consider what to surround according to what can be considered as content inside tags. Consider the following:
```html
<div>
  <label for="test">test</label>
  <input id="test" />
</div>
```
If you put your cursor on the label tag, it will surround around the label tag since the text between the opening and closing tags is considered the content. Whereas if you put your cursor on the input tag, since it doesn't have any content (empty tag) it will surround all the way out to and including the div tags, since it considers everything inbetween as content.

**Reply by ShooTeX** [reply]
sorry the video is a little confusing, at the beginning I have the cursor on div, and the closing tag is correctly highlighted. trying `vat` selects the parent div (not expected) and a closing tag that doesn't belong to it. It works like that for some other elements (elements that don't have self-closing elements inside, maybe?)

**Reply by ShooTeX** [reply]
I think in this video it's easier to see:

https://user-images.githubusercontent.com/10850738/228384862-f21dc219-e55c-457c-b4d5-366e68df74a7.mov

**Reply by dpetka2001** [reply]
Your second video is not working, so I'm going to respond according to what I saw in the first (maybe they're deleted after some time?). Let's say that `*` will denote your cursor's position. In your first video, your cursor was inside a `<*div>` element. Since there is no immediate content to consider (like I said in my example above with the different behavior between `input` and `label`), it considers that where your current cursor position is to be the content of the parent tag (which is the outside div element). So it chooses all tags inside the parent element as content along with the outside parent tags. Maybe you would like to make an issue at `mini.ai`'s repo to address the issue of what can be considered content.

**Reply by dpetka2001** [reply]
Sorry for the confusion I tried to see the video when I was replying and didn't show anything. The video is working fine, but my answer still stands. In your second video it correctly wraps around in the first try, because everything inside your `<div>` is considered as content. In the second try where there's just the outside `<div>` element, it considers everything as content and wraps around the `form` element in the end. That's because there's no specific content inside the outside `<div>` element. There are just 2 nested div elements, with the first div element not having explicit content. Again, you might want to make an issue at `mini.ai`'s repo to discuss the boundaries of what can be considered content (implicitly or explicitly).

**Reply by ShooTeX** [reply]
I don't think I understand what you mean by _no specific_ content, because the content of the parent div is `div` and `div`, while the content of the child div is `Label` and `Input`. But I will try making an issue in the mini repo, thanks!

https://github.com/echasnovski/mini.nvim/discussions/285

**Reply by dpetka2001** [reply]
Yeah, I don't think that it's considering nested divs as content (based on its behavior). I think it's explicitly looking for content inside the opening/closing tags. So, from its point of operation, 
```html
<div>
  <div>
    <label for="test">test2</label>
    <input id="test" />
  </div>
</div>
```
The first div has content div, so if your cursor is inside the second div, then it will visually select everything. If your cursor is at label, which has explicit content between its tags, it will visually select around `label`. If your cursor is at `input`, which doesn't have any explicit content, then it will visually select around the inside div, because since there is no explicit content at the location of your cursor, it considers everything as content to the parent element (in this case the inside `div` element). I'm sorry if my English confuses you, since it's not my native language and maybe I don't express myself as accurately as I should. But the creator of the addon will surely be more appropriate to solve your questions for sure.
